File: lib/domain/cooldown.dart
enum CooldownAction {
  /// Open without a warning dialog.
  allow,

  /// Show the yes/no warning dialog.
  warn,
}

/// Pure cooldown rules shared as the spec for the Android native monitor.
class Cooldown {
  const Cooldown._();

  static const defaultPeriod = Duration(hours: 1);
  static const closeGrace = Duration(seconds: 3);

  static const presets = <Duration>[
    Duration(minutes: 15),
    Duration(minutes: 30),
    Duration(hours: 1),
    Duration(hours: 3),
    Duration(hours: 6),
    Duration(hours: 24),
  ];

  static CooldownAction decide({
    required bool watched,
    required bool enabled,
    required bool sessionAllowed,
    required DateTime? lastClosedAt,
    required Duration warningPeriod,
    required DateTime now,
  }) {
    if (!watched || !enabled) {
      return CooldownAction.allow;
    }
    if (sessionAllowed) {
      return CooldownAction.allow;
    }
    if (lastClosedAt == null) {
      return CooldownAction.allow;
    }
    if (now.difference(lastClosedAt) < warningPeriod) {
      return CooldownAction.warn;
    }
    return CooldownAction.allow;
  }

  static Duration? remainingCooldown({
    required DateTime? lastClosedAt,
    required Duration warningPeriod,
    required DateTime now,
  }) {
    if (lastClosedAt == null) {
      return null;
    }
    final until = lastClosedAt.add(warningPeriod);
    if (!until.isAfter(now)) {
      return null;
    }
    return until.difference(now);
  }
}

File: android/app/src/main/kotlin/com/komapotter/miruna/monitor/Cooldown.kt
package com.komapotter.miruna.monitor

object Cooldown {
    const val DEFAULT_PERIOD_MS = 60L * 60L * 1000L
    const val CLOSE_GRACE_MS = 3_000L

    const val ACTION_ALLOW = "allow"
    const val ACTION_WARN = "warn"

    fun decide(
        watched: Boolean,
        enabled: Boolean,
        sessionAllowed: Boolean,
        lastClosedAtMs: Long?,
        warningPeriodMs: Long,
        nowMs: Long,
    ): String {
        if (!watched || !enabled) return ACTION_ALLOW
        if (sessionAllowed) return ACTION_ALLOW
        if (lastClosedAtMs == null) return ACTION_ALLOW
        if (nowMs - lastClosedAtMs < warningPeriodMs) return ACTION_WARN
        return ACTION_ALLOW
    }

    fun shouldRecordClose(unconfirmedWarning: Boolean): Boolean = !unconfirmedWarning

    fun formatWarningPeriod(ms: Long): String {
        val totalMinutes = (ms / 60_000L).coerceAtLeast(0)
        val hours = totalMinutes / 60
        val minutes = totalMinutes % 60
        return when {
            hours > 0 && minutes > 0 -> "${hours}時間${minutes}分"
            hours > 0 -> "${hours}時間"
            else -> "${minutes}分"
        }
    }
}
